src/App.tsx

```diff
@@ -1,24 +1,214 @@
+      switch (todoStatus) {
+        case TodoStatus.ACTIVE:
+          return !todo.completed;
+        case TodoStatus.COMPLETED:
+          return todo.completed;
+        default:
+          return true;
+      }
+    })
+    );
+  }, [todoStatus, todos]);
+
+  const fetchData = async () => {
+    try {
+      const todosFromServer = await getTodos(USER_ID);
+
+      setTodos(todosFromServer);
+    } catch (error) {
+      setErrorMessage(ErrorMessage.LOAD);
+    }
+  };
+
+  useEffect(() => {
+    fetchData();
+  }, []);
+
+  const addNewTodo = async (todoTitle: string) => {
+    try {
+      const newTodo: TodoToSend = {
+        userId: USER_ID,
+        title: todoTitle,
+        completed: false,
+      };
+
+      setTempoTodo({
+        id: 0,
+        ...newTodo,
+      });
+      const addedTodo = await addTodos(newTodo);
+
+      setTodos(currentTodos => [...currentTodos, addedTodo]);
+    } catch (error) {
+      setErrorMessage(ErrorMessage.ADD);
+    } finally {
+      setTempoTodo(null);
+    }
+  };
+
+  const removeTodo = async (id: number) => {
+    try {
+      setLoadingTodo(prev => [...prev, id]);
+      await deleteTodos(id);
+
+      setTodos(prevTodos => prevTodos
+        .filter(todo => todo.id !== id));
+    } catch (error) {
+      setErrorMessage(ErrorMessage.DELETE);
+    } finally {
+      setTempoTodo(null);
+      setLoadingTodo([0]);
+    }
+  };
+
+  const updateTitleOfTodo = async (updatingTodo: Todo, title?: string) => {
+    setLoadingTodo(prev => [...prev, updatingTodo.id]);
+
+    const updatedTodo = title
+      ? ({ ...updatingTodo, title })
+      : ({ ...updatingTodo, completed: !updatingTodo.completed });
+
+    try {
+      await updateTodoTitle(updatedTodo);
+      await fetchData();
+    } catch {
+      setErrorMessage(ErrorMessage.UPDATE);
+    } finally {
+      setLoadingTodo(
+        current => current.filter(todoId => todoId !== updatingTodo.id),
+      );
+    }
+  };
+
+  const updateStatusOfTodo = async (updatingTodo: Todo) => {
+    const newStatus = !updatingTodo.completed;
+
+    try {
+      setLoadingTodo(current => [...current, updatingTodo.id]);
+
+      const updatedTodo: Todo = await updateTodoStatus(
+        updatingTodo.id,
+        newStatus,
+      );
+
+      setTodos(currentTodos => (
+        currentTodos.map(todo => (
+          todo.id === updatingTodo.id
+            ? updatedTodo
+            : todo
+        ))
+      ));
+    } catch {
+      setErrorMessage(ErrorMessage.UPDATE);
+    } finally {
+      setLoadingTodo(
+        current => current.filter(todoId => todoId !== updatingTodo.id),
+      );
+    }
+  };
+
+  const toggleAll = () => {
+    let toggledTodos = todos;
+
+    if (activeTodosCount) {
+      toggledTodos = todos.filter(todo => !todo.completed);
+    }
+
+    toggledTodos.forEach(todo => {
+      updateStatusOfTodo(todo);
+    });
+  };
+
+  const clearCompletedTodos = () => {
+    todos.forEach(async (todo) => {
```

Comment:
Consider using `Promise.all` and map todos to promises (i.e just remove async and await keywords) for better performance